src/software/backend/wifi_backend.cpp

```diff
@@ -0,0 +1,74 @@
+#include "software/backend/wifi_backend.h"
+
+#include "shared/constants.h"
+#include "software/constants.h"
+#include "software/parameter/dynamic_parameters.h"
+#include "software/proto/message_translation/protobuf_message_translator.h"
+#include "software/util/design_patterns/generic_factory.h"
+
+const std::string WifiBackend::name = "wifi";
+
+WifiBackend::WifiBackend()
+    : network_input(Util::Constants::SSL_VISION_DEFAULT_MULTICAST_ADDRESS,
+                    Util::Constants::SSL_VISION_MULTICAST_PORT,
+                    Util::Constants::SSL_GAMECONTROLLER_MULTICAST_ADDRESS,
+                    Util::Constants::SSL_GAMECONTROLLER_MULTICAST_PORT,
+                    boost::bind(&WifiBackend::receiveWorld, this, _1),
+                    Util::DynamicParameters->getAIControlConfig()->getRefboxConfig(),
+                    Util::DynamicParameters->getCameraConfig()),
+      io_service()
+{
+    // connect to current channel
+    joinMulticastChannel(Util::DynamicParameters->getNetworkConfig()->Channel()->value());
+
+    // TODO can't register callbacks with the immutable parameters, why?
+    // add callback to switch channels on param change
+    Util::MutableDynamicParameters->getMutableNetworkConfig()
+        ->mutableChannel()
+        ->registerCallbackFunction(
+            boost::bind(&WifiBackend::joinMulticastChannel, this, _1));
+
+    // start the thread to run the io_service in the background
+    io_service_thread = std::thread([this]() { io_service.run(); });
+}
+
+WifiBackend::~WifiBackend()
+{
+    io_service_thread.join();
+}
+
+void WifiBackend::onValueReceived(ConstPrimitiveVectorPtr primitives_ptr)
+{
+    primitive_output->sendProto(
+        *ProtobufMessageTranslator::getPrimitiveMsgFromPrimitiveVector(primitives_ptr));
+}
+
+void WifiBackend::receiveWorld(World world)
+{
+    vision_output->sendProto(*ProtobufMessageTranslator::getVisionMsgFromWorld(world));
+    Subject<World>::sendValueToObservers(world);
+}
+
+void WifiBackend::receiveTbotsRobotMsg(TbotsRobotMsg robot_msg)
+{
+    SensorMsg sensor_msg;
+    TbotsRobotMsg* added_robot_msg = sensor_msg.add_tbots_robot_msg();
+    *added_robot_msg               = robot_msg;
```

Comment:
We should standardize this with how we do unique ptrs. If we decide on the unique ptr way, then this should be changed to that.

Reply:
I think we ended up deciding on the unique ptr way?

Reply:
I tried this and its not possible to do the unique_ptr approach unfortunately, the callbacks can't be defined like this:

`std::function<void(std::unique_ptr<ReceiveProto>)> receive_callback)`

thoughts on leaving it as is?

Reply:
the other alternative to avoid copying values would be shared ptr but I'm not sure how ppl feel about that

Reply:
This won't work with `shared_ptrs` because `shared_ptrs` don't have `.release()` IIRC

Reply:
@akhilveeraghanta have you tried passing `unique_ptr` by reference or using `std::move`?

Reply:
Curious to see whether that works. It would need to be copy-constructable https://stackoverflow.com/questions/31691309/how-to-return-a-stdfunction-that-contains-a-stdunique-ptr-from-a-generalized

Reply:
`std::move` doesn't work that was what I tried, it complains about r-value and l-value stuff when u bind the function. And I was generally avoiding passing the smart ptr by reference because that is usually frowned upon.

We are using unique_ptrs to remove raw ptrs during protobuf message conversion. Im inclined to leave this as is to avoid doing something that adds complexity or goes against the standards.

Reply:
I think we should either use unique_ptrs everywhere or use them nowhere. We can discuss this offline if you'd like